compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2332,14 +2332,105 @@ impl<'a> Parser<'a> {
+                // the AST for typechecking.
+                err.span_label(ident_span, "while parsing this `fn`");
+                err.emit();
+                Ok(())
+            } else if self.token.kind == token::RArrow
+                && let Some(fn_params_end) = fn_params_end
+            {
+                // Instead of a function body, the parser has encountered a right arrow
+                // preceded by a where clause.
+
+                // Find whether token behind the right arrow is a function trait and
+                // store its span.
+                let fn_trait_span =
+                    [sym::FnOnce, sym::FnMut, sym::Fn].into_iter().find_map(|symbol| {
+                        if self.prev_token.is_ident_named(symbol) {
+                            Some(self.prev_token.span)
+                        } else {
+                            None
+                        }
+                    });
+
+                // Parse the return type (along with the right arrow) and store its span.
+                // If there's a parse error, cancel it and return the existing error
+                // as we are primarily concerned with the
+                // expected-function-body-but-found-something-else error here.
+                let arrow_span = self.token.span;
+                let ty_span = match self.parse_ret_ty(
+                    AllowPlus::Yes,
+                    RecoverQPath::Yes,
+                    RecoverReturnSign::Yes,
+                ) {
+                    Ok(ty_span) => ty_span.span().shrink_to_hi(),
+                    Err(parse_error) => {
+                        parse_error.cancel();
+                        return Err(err);
+                    }
+                };
+                let ret_ty_span = arrow_span.to(ty_span);
+
+                if let Some(fn_trait_span) = fn_trait_span {
+                    // Typo'd Fn* trait bounds such as
+                    // fn foo<F>() where F: FnOnce -> () {}
+                    err.subdiagnostic(errors::FnTraitMissingParen { span: fn_trait_span });
+                } else if let Ok(snippet) = self.psess.source_map().span_to_snippet(ret_ty_span) {
+                    // If token behind right arrow is not a Fn* trait, the programmer
+                    // probably misplaced the return type after the where clause like
+                    // `fn foo<T>() where T: Default -> u8 {}`
+                    err.primary_message(
+                        "return type should be specified after the function parameters",
+                    );
+                    err.subdiagnostic(errors::MisplacedReturnType {
+                        fn_params_end,
+                        snippet,
+                        ret_ty_span,
+                    });
+                }
+                Err(err)
+            } else {
+                Err(err)
+            }
+        } else {
+            Ok(())
```

Comment:
that `Ok(())` is very questionable :thinking: I would expect that this is unreachable?

Reply:
Replacing the `Ok(())` with `unreachable!()` fails on this UI test:
``` rust
extern "C" {
  fn foo() //~ERROR expected `;`
}
```
It's because `expected_one_of_not_found()` returns an `Ok()` when it's able to recover from an error.

Reply:
:+1: two more requests in that case. `expected_one_of_not_found` currently always returns `Recovered::Yes(ErrorGuaranteed)` in that case afaict. Please change it's return type to `PResult<'a, ErrorGuaranteed>` and then use a match with `Err::<_; ErrorGuaranteed>(_guar) => Ok(()),` or sth to make sure we only return `Ok(())` without an error if we've already emitted one

Reply:
Sure, changed the return type of those two functions to `PResult<'a, ErrorGuaranteed>`. Also, made some small changes to `expect_one_of` to map `ErrorGuaranteed` to `Recovered::Yes(ErrorGuaranteed)`.

Thank you for the suggestions.